lstchain/calib/camera/r0.py

```diff
@@ -0,0 +1,288 @@
+                            pixel = expected_pixel_id[nr_clus*7 + pix]
+                            interpolate_spike_A(waveform, gain, spike_A_position, pixel)
+
+                    spike_b_position = int(
+                        (fc_old[nr_clus, gain, pix] - 1 - fc[nr_clus, gain, pix] + 2 * size4drs) % size4drs)
+                    if spike_b_position < roisize - 1:
+                        pixel = expected_pixel_id[nr_clus*7 + pix]
+                        interpolate_spike_B(waveform, gain, spike_b_position, pixel)
+        return waveform
+
+    def _load_calib(self):
+        """event.r0.tel[0].waveform
+        If a pedestal file has been supplied, create a array with
+        pedestal value . If it hasn't then point calibrate to
+        fake_calibrate, where nothing is done to the waveform.
+        """
+        if self.pedestal_path:
+            with fits.open(self.pedestal_path) as f:
+                n_modules = f[1].header['NAXIS4']
+                self.pedestal_value_array = np.zeros((n_modules, 2, 7, 4136), dtype=np.int16)
+                pedestal_data = np.int16(f[1].data)
+                self.pedestal_value_array[:, :, :, :self.size4drs] = pedestal_data - self.offset
+                self.pedestal_value_array[:, :, :, self.size4drs:self.size4drs + 40] \
+                    = pedestal_data[:, :, :, 0:40] - self.offset
+
+    def _get_first_capacitor(self, event, nr_module):
+        """
+        Get first capacitor values from event for nr module.
+        Parameters
+        ----------
+        event : `ctapipe` event-container
+        nr_module : number of module
+        """
+        fc = np.zeros((2, 7))
+        first_cap = event.lst.tel[0].evt.first_capacitor_id[nr_module * 8:
+                                                            (nr_module + 1) * 8]
+        for i, j in zip([0, 1, 2, 3, 4, 5, 6], [0, 0, 1, 1, 2, 2, 3]):
+            fc[self.high_gain, i] = first_cap[j]
+        for i, j in zip([0, 1, 2, 3, 4, 5, 6], [4, 4, 5, 5, 6, 6, 7]):
+            fc[self.low_gain, i] = first_cap[j]
+        return fc
+
+@jit(parallel=True)
+def subtract_pedestal_jit(event_waveform, expected_pixel_id, fc_cap, pedestal_value_array, n_modules):
+    waveform = np.zeros(event_waveform.shape)
+    size4drs = 4096
+    n_gain = 2
+    n_pix = 7
+    for nr in prange(0, n_modules):
+        for gain in prange(0, n_gain):
+            for pix in prange(0, n_pix):
+                pixel = expected_pixel_id[nr*7 + pix]
+                position = int((fc_cap[nr, gain, pix]) % size4drs)
+                waveform[gain, pixel, :] = \
+                    (event_waveform[gain, pixel, :] -
+                    pedestal_value_array[nr, gain, pix, position:position + 40])
+    return waveform
+
+@jit(parallel=True)
+def do_time_lapse_corr(waveform, expected_pixel_id, local_clock_list, fc, last_time_array, number_of_modules):
+    size4drs = 4096
+    for nr_clus in prange(0, number_of_modules):
```

Comment:
All these for loops are likely making the code slower, would there be a more pythonic way of doing it using numpy functionalities instead of going though all the loops?

Reply:
I improved code a bit. But now, I would like leave code as simple as possible, because in next step I will modify code to take account another small effects.